src/connector/src/sink/mqtt.rs

```diff
@@ -242,17 +379,65 @@ struct MqttSinkPayloadWriter {
     config: MqttConfig,
     qos: QoS,
     retain: bool,
+    path: Vec<usize>,
 }
 
-impl FormattedSink for MqttSinkPayloadWriter {
-    type K = Vec<u8>;
-    type V = Vec<u8>;
+impl MqttSinkPayloadWriter {}
+
+impl MqttSinkPayloadWriter {
+    async fn write_chunk(&mut self, chunk: StreamChunk, encoder: &RowEncoderWrapper) -> Result<()> {
+        for (op, row) in chunk.rows() {
+            if op != Op::Insert {
+                continue;
+            }
+
+            let topic = if self.path.is_empty() {
+                self.config.common.topic.as_str()
+            } else {
+                let mut iter = self.path.iter();
+
+                let scalar = iter
+                    .next()
+                    .and_then(|pos| row.datum_at(*pos))
+                    .and_then(|d| {
+                        iter.try_fold(d, |d, pos| match d {
+                            ScalarRefImpl::Struct(struct_ref) => {
+                                struct_ref.iter_fields_ref().nth(*pos).flatten()
+                            }
+                            _ => None,
+                        })
+                    });
+                match scalar {
+                    Some(ScalarRefImpl::Utf8(s)) => s,
+                    _ => {
+                        if self.config.fallback_to_default {
+                            self.config.common.topic.as_str()
+                        } else {
+                            return Err(SinkError::Mqtt(anyhow!(
```

Comment:
*Continued from https://github.com/risingwavelabs/risingwave/pull/15673#discussion_r1530621620*

I have a different opinion.

It's unacceptable to throw an error when processing data. Once an error happens, the streaming job will recover to last checkpoint, and the same data comes in, causing error again... Users can't do anything, but only feel panicked.

In my mind, when running into a row with NULL value in `topic`

- If a default topic is specified, fallback to it silently.
- Otherwise, print an `ERROR` log telling the users a row is dropped because of missing of topic.

Reply:
@fuyufjh  Silently failing can be confusing too, and harder for the users to debug, one of the aspects I think it still needs to be improved in RisingWave is error management, things like what errors are retriable, the number of retries, backoffs, dead letter queues are missing. In the meantime I implemented these changes